### PR TITLE
Add ADR-004-QGB_relayer_security

### PR DESCRIPTION
**Pull request by SweeXordious**
ADR for the relayer security issue we have.

### CODE REVIEW

**Review by evan-forbes — COMMENTED** [review]
This is a pretty good start, and after we expand on it a bit we should be good to go as far as beginning the design process.

File: docs/architecture/ADR-004-QGB_relayer_security.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+Also, a malicious relayer can create the above issue in any QGB bridge:  
+-   Get the latest relayed valset
+-   Listen for new signed valsets
+-   Once a new valset is signed by 2/3 of the network, submit it immediately to be included in next block
+
+If the honest relayer was slow in submitting the data commitments that were signed by the previous valset, then the bridge will have data commitment signatures holes. Because, the honest relayer will no longer be able to submit the previous data commitments as they're not, in general, signed by the valset relayed by the malicious relayer. And theonly solution is to jump to the ones that were signed with the current valset.
+
+## Alternative Approaches
+
+### Catchup solutions
```
[the remark this answers is not in the document]
Comment (SweeXordious) [reply docs/architecture/ADR-004-QGB_relayer_security.md:35]: removed

File: docs/architecture/ADR-004-QGB_relayer_security.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+#### A request oriented design
+Currently, the attestations that need to be signed are defined by the state machine based on `abci.endBlock()` and a `DataCommitmentWindow`. This simplifies the state machine and doesn't require  implementing new transaction types to ask for orchestrators signatures.
+
+The request oriented design means providing users (relayers mainly) with the ability to post data commitment requests and ask orchestrators to sign them.
+
+To avoid spamming the network with requests, these signatures can be added to the block data in specific namespaces and the we can charge per request.
+
+- Pros
+	- Would fix the relayer issue described above
```
[the remark this answers is not in the document]
Comment (SweeXordious) [reply docs/architecture/ADR-004-QGB_relayer_security.md:72]: true :smile:

File: docs/architecture/ADR-004-QGB_relayer_security.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+- {date}: {changelog}
+
+## Context
+
+Currently, the QGB  smart contract is designed to update the data commitments as follows:  
+- Receive a data commitment  
+- Check that the block height (nonce) is  higher than the previously committed root  
+- Check if the data commitment is signed using the current valset _(this is the problematic check)_  
+- Then, other checks + commit
+So, if a relayer is up to date, it will submit data commitment and will pass the above checks.
```
[the remark this answers is not in the document]
Comment (SweeXordious) [reply docs/architecture/ADR-004-QGB_relayer_security.md:15]: check now please?

File: docs/architecture/ADR-004-QGB_relayer_security.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+- Pros:  
+		- Would make the relayer parallelizable (can submit data commitments and valsets in any order as long as the valset is committed)  
+		- would allow the QGB to catchup correctly even in the existence of a malicious relayer  
+
+#### A request oriented design
+Currently, the attestations that need to be signed are defined by the state machine based on `abci.endBlock()` and a `DataCommitmentWindow`. This simplifies the state machine and doesn't require  implementing new transaction types to ask for orchestrators signatures.
+
+The request oriented design means providing users (relayers mainly) with the ability to post data commitment requests and ask orchestrators to sign them.
+
+To avoid spamming the network with requests, these signatures can be added to the block data in specific namespaces and the we can charge per request.
```
[the remark this answers is not in the document]
Comment (SweeXordious) [reply docs/architecture/ADR-004-QGB_relayer_security.md:56]: check now please?

File: docs/architecture/ADR-004-QGB_relayer_security.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+- Pros
+	- Would fix the relayer issue described above
+	- Would allow anyone to ask for signatures over commitments, ie, the QGB can then be used by any team without changing anything.
+
+- Cons
+	- Makes slashing more complicated
```
[the remark this answers is not in the document]
Comment (SweeXordious) [reply docs/architecture/ADR-004-QGB_relayer_security.md]: check now please?

File: docs/architecture/ADR-004-QGB_relayer_security.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+Then, this would create holes in the signatures as the honest relayer will no longer be able to submit the previous data commitments as they're not, in general, signed by the valset relayed by the malicious relayer. And the only solution is to jump to the ones that were signed with the current valset.
+
+## Alternative Approaches
+
+###  More  synchrony : Deploy the QGB contract with a data commitment window
+When deploying the QGB  contract,  also  set the data commitment window,  ie, the number of blocks between the `beginBlock` and `endBlock` of each data  commitment confirm.
+
+Then, update the QGB contract to check when receiving a new valset if the latest relayed data commitment height is >= new valset height - data commitment window.
+
+This also would mean adding, for example, a `DataCommitmentWindowConfirm` representing signatures of the validator set for a certain `DataCommitmentWindow`, since this latter can be updated using gov proposals.
```
Comment (evan-forbes) [inline docs/architecture/ADR-004-QGB_relayer_security.md:36]: we probably want to update this portion of the ADR now to reflect our current plan to use universal nonce, but I like the wording here that describes the goal of adding synchrony. We have a data commitment window but the contract *is not* aware of it, which is an issue. How do we plan on using universal nonces, why does this fix the data commitment issue, and how does this fit in with the security assumptions the contract is already making?

File: docs/architecture/ADR-004-QGB_relayer_security.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+- Pros
+	- Would allow anyone to ask for signatures over commitments, ie, the QGB can then be used by any team without changing anything.
+
+- Cons
+	- Makes slashing more complicated. In fact, to slash for liveness, providing the whole history of blocks, proving that a certain orchestrator didn't sign an attestation in the given period, will be hard to implement and the proofs will be big. Compared to the attestations being part of the state, which can be queried easilly.
+
+## Decision
+
+Still not made.
```
Comment (evan-forbes) [inline docs/architecture/ADR-004-QGB_relayer_security.md:68]: barring something popping up, I think we can update this section now too

File: docs/architecture/ADR-004-QGB_relayer_security.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+Still not specified.
+
+## Status
+
+Proposed
+
+## References
+
+> Are there any relevant PR comments, issues that led up to this, or articles referenced for why we made the given design choice? If so link them here!
```
Comment (evan-forbes) [inline docs/architecture/ADR-004-QGB_relayer_security.md:80]: note to self that we should add the contracts PR here if its up before we merge this, if not tho nbd

**Review by evan-forbes — COMMENTED** [review]
:+1: after we update this with decision and detailed design after the contracts PR is up, this ADR is good to go!

File: docs/architecture/ADR-004-QGB_relayer_security.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (evan-forbes) [reply docs/architecture/ADR-004-QGB_relayer_security.md]: perfect!

File: docs/architecture/ADR-004-QGB_relayer_security.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (evan-forbes) [reply docs/architecture/ADR-004-QGB_relayer_security.md]: :+1: :100:

File: docs/architecture/ADR-004-QGB_relayer_security.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+## Context
+
+Currently, the QGB  smart contract is designed to update the data commitments as follows:  
+- Receive a data commitment  
+- Check that the block height (nonce) is  higher than the previously committed root  
+- Check if the data commitment is signed using the current valset _(this is the problematic check)_  
+- Then, other checks + commit
+So, if a relayer is up to date, it will submit data commitment and will pass the above checks.
+
+Now, if the relayer is missing some data commitments or valset updates, then it will start catching up. This will happen at least one time, as we will not launch the network with the QGB, but it will be added through an upgrade.
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply docs/architecture/ADR-004-QGB_relayer_security.md]: :+1:

File: docs/architecture/ADR-004-QGB_relayer_security.md (hunk 6 above)
Comment (SweeXordious) [reply docs/architecture/ADR-004-QGB_relayer_security.md:43]: check now please? I added another section for the synchronous approach

File: docs/architecture/ADR-004-QGB_relayer_security.md (hunk 7 above)
Comment (SweeXordious) [reply docs/architecture/ADR-004-QGB_relayer_security.md]: check now please?

File: docs/architecture/ADR-004-QGB_relayer_security.md (hunk 8 above)
Comment (SweeXordious) [reply docs/architecture/ADR-004-QGB_relayer_security.md]: check now please?

File: docs/architecture/ADR-004-QGB_relayer_security.md [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+the **Synchronous QGB : Universal nonce approach** will be implemented as it will allow us to ship a working QGB 1.0 version faster while preserving the same security assumptions at the expense of parallelization, and custumization, as discussed under the _request oriented design_ above.
+
+## Detailed Design
+
+The implementation will mainly consist of:
+
+- Updating the state machine to output valset requests with the correct nonce
+- Add state machine tests
+- Update the orchestrator to generate the right nonce for data commitments 
+- Update the relayer implementation to be sequential
```
Comment (evan-forbes) [inline docs/architecture/ADR-004-QGB_relayer_security.md:96]: Considering the the design changes that we went with is relatively small compared to some of the other options, I'm perfectly fine with not filling in all of these details now. In fact, I think this creates a fantastic opportunity for a future contributor #471 to dig through the code and figure out how it actually works.

**Review by evan-forbes — APPROVED** [review]
nice work! I think we can merge this and focus on the implementation now.

File: docs/architecture/ADR-004-QGB_relayer_security.md (hunk 10 above)
Comment (SweeXordious) [inline docs/architecture/ADR-004-QGB_relayer_security.md:96]:
```suggestion
```

**Pull request merged**